R/DataSetList.R

```diff
@@ -700,48 +730,165 @@ get_runtimes <- function(dsList) {
-# }
+#' Get all options for a specific attribute which can be used to subset a DataSetList
+#' 
+#' This is a more generic version of the existing `get_dim`, `get_funcId` and `get_algId` functions.
+#' Note the only attributes returned by `get_static_attributes` are supported in this funcion
+#'
+#' @param dsl The DataSetList
+#' @param attribute the name of the attribute for which to get the available options in dsl
+#' @return The list of options for the specified attribute
+#' @export
+#' @examples
+#' get_static_attribute_values(dsl, 'funcId')
+get_static_attribute_values <- function(dsl, attribute) {
+  unique(unlist(lapply(dsl, function(ds) {attr(ds, attribute)})))
+}
 
-# TODO: the attribute list should als be sliced here...
 #' Filter a DataSetList by some criteria
 #'
-#' @param x The DataSetLsit
-#' @param ... The condition to filter on. Can be any expression which assigns True or False
-#' to a DataSet object, such as DIM == 625 or funcId == 2
+#' @param x The DataSetList
+#' @param ... The conditions to filter on. Can be any expression which assigns True or False
+#' to a DataSet object, such as DIM == 625 or funcId == 2. Usage of && and || is only supported on default attributes 
+#' (funcId, algId, DIM), not on combinations of with other attributes (e.g. instance). In those cases, & and | should 
+#' be used respectively. Alternatively, this can be used as a keyword argument named 'text', with the condition as a 
+#' string to be parsed. This allows exectution of subset commands on arbitrary variables in code. 
 #'
 #' @return The filtered DataSetList
 #' @export
 #' @examples
 #' subset(dsl, funcId == 1)
+#' subset(dsl, funcId == 1 && DIM == 16) # Can use && and || for default attributes
+#' subset(dsl, instance == 1)
+#' subset(dsl, instance == 1 & funcId == 1) # Can use & and | for all attributes
+#' subset(dsl, instance == 1, funcId == 1) # Comma-seperated conditions are treated as AND
 subset.DataSetList <- function(x, ...) {
-  condition_call <- substitute(list(...))
   enclos <- parent.frame()
-  idx <- sapply(x,
-                function(ds)
-                  all(unlist(
-                    eval(condition_call, attributes(ds), enclos)
-                  )))
-  x[idx]
+  if (hasArg('text')) {
+    text <- list(...)$text
+    condition_call <- parse(text = text)
+  } else {
+    condition_call <- substitute(list(...))
+    condition_call <- condition_call[2:length(condition_call)]
+  }
+  
+  obj <- lapply(x,
+                function(ds){
+                  mask <- tryCatch(expr = {
+                    mask <- NULL
+                    for (idx in seq(length(condition_call))) {
+                      mask_temp <- unlist(
+                        eval(condition_call[[idx]], attributes(ds), enclos = enclos)
+                      )
+                      if (is.null(mask)) mask <- mask_temp
+                      else {
+                        if (length(mask_temp) == 1 && !mask_temp) {
+                          mask <- F
```

Comment:
Should we exit the loop from here?  I think maybe we could make this loop simpler by force the `mask_temp` to always have length of the runs in a dataset? then you just need to take bitwise `and` in the loop?